Adding elements to code coverage config passed via commandline,

fhnaseer opened the pull request:
Adding elements to code coverage config passed via commandline,

e.g.
Input:
`--collect:"Code Coverage;CLRIEInstrumentationNetCore=true;CLRIEInstrumentationNetFramework=false"`

Output:
```
      <DataCollector friendlyName="Code Coverage" enabled="True">
        <Configuration>
          <CLRIEInstrumentationNetCore>true</CLRIEInstrumentationNetCore>
          <CLRIEInstrumentationNetFramework>false</CLRIEInstrumentationNetFramework>
        </Configuration>
      </DataCollector>
```

fhnaseer committed:
Adding elements to code coverage config passed via commandline,

e.g.
Input:
--collect:"Code Coverage;CLRIEInstrumentationNetCore=true;CLRIEInstrumentationNetFramework=false"

Output:
      <DataCollector friendlyName="Code Coverage" enabled="True">
        <Configuration>
          <CLRIEInstrumentationNetCore>true</CLRIEInstrumentationNetCore>
          <CLRIEInstrumentationNetFramework>false</CLRIEInstrumentationNetFramework>
        </Configuration>
      </DataCollector>

fhnaseer committed:
Adding tests,

jakubch1 commented:
Please add acceptance tests for code coverage which will cover Format setting.
If runsettings are also specified, will this overwrite value in runsettings?

jakubch1 commented on test/vstest.console.UnitTests/Processors/CollectArgumentProcessorTests.cs:
```diff
@@ -586,6 +585,80 @@ public void InitializeXPlatCodeCoverageShouldAddXPlatInProcProcDataCollectoPrope
+                "    <DataCollectors>",
+                "      <DataCollector friendlyName=\"MyDataCollector\" enabled=\"True\">",
+                "        <Configuration>",
+                "          <SomeSetting>SomeValue</SomeSetting>",
```
please add test also when we overwrite value from runsettings

fhnaseer replied:
Added.

jakubch1 commented on test/vstest.console.UnitTests/Processors/CollectArgumentProcessorTests.cs:
```diff
@@ -586,6 +585,80 @@ public void InitializeXPlatCodeCoverageShouldAddXPlatInProcProcDataCollectoPrope
+                "      </DataCollector>",
+                "    </DataCollectors>",
+                "  </DataCollectionRunSettings>",
+                "</RunSettings>"), this.settingsProvider.ActiveRunSettings.SettingsXml);
```
Add also some testing in case you have 2 collectors enabled:

--collect "datacollector1;arg1=val1,arg2=val2;datacollector2;arg3=val3,arg4=val4"

fhnaseer replied:
In previous implementation `dotnet test --collect "Code Coverage;XPlat Code Coverage"` doesn't add two collectors. It will add only 1 with friendly name "Code Coverage;XPlat Code Coverage"
```
    <DataCollectors>
      <DataCollector friendlyName="MyDataCollector;AnotherCollector" enabled="True" />
    </DataCollectors>
```

jakubch1 replied:
You are basing it probably on unit tests results but this class that you are testing is getting string which was already splitted by ";". This is why you need to add acceptance tests :). Another thing that there is already special processor for "Code Coverage".

Check this:
```
D:\examples\xunit170> dotnet test --collect "Code Coverage;XPlat Code Coverage"
  Determining projects to restore...
  All projects are up-to-date for restore.
  xunit170 -> D:\examples\xunit170\bin\Debug\net6.0\xunit170.dll
Test run for D:\examples\xunit170\bin\Debug\net6.0\xunit170.dll (.NETCoreApp,Version=v6.0)
Microsoft (R) Test Execution Command Line Tool Version 17.0.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
A total of 1 test files matched the specified pattern.

Attachments:
  D:\examples\xunit170\TestResults\07f5cdaa-e532-43a0-a96a-4533eba6df87\coverage.cobertura.xml
  D:\examples\xunit170\TestResults\07f5cdaa-e532-43a0-a96a-4533eba6df87\UseThisAccount_JACHOCHO-5820_2021-11-11.12_22_46.coverage
Passed!  - Failed:     0, Passed:     1, Skipped:     0, Total:     1, Duration: < 1 ms - xunit170.dll (net6.0)
```

fhnaseer replied:
From code, we can add multiple `collect` commands. So to specify multiple collectors we can use this.
`dotnet text --collect:"Collector1" --collect:"Collector2"`
With configurations.
`dotnet text --collect:"Collector1;arg1=val1" --collect:"Collector2;arg2=val2"`

fhnaseer replied:
Acceptance tests accepts runsettings string and doesn't take command line arguments. @nohwnd Do you know if there are any acceptance tests which takes both runsettings and command line arguments?

jakubch1 replied:
what if someone is already using --collect "datacollector1;datacollector2"? Your code would throw exception. I think we shouldn't make such breaking change

nohwnd replied:
@jakubch1 We did not find such usage on github, and we can't say for sure, but I agree to throw error in that case. Logger parameter does the same. 

@fhnaseer ExecuteTestsWithDataCollection seems to be providing both runsettings as well as command line parameters. or is that not what you want?  

https://github.com/microsoft/vstest/blob/ad32654bc155a059fda263f0ec99b445058d84ea/test/Microsoft.TestPlatform.AcceptanceTests/DataCollectionTests.cs#L36-L37

fhnaseer replied:
Added acceptance tests in test/Microsoft.TestPlatform.AcceptanceTests/CodeCoverageTests.cs

jakubch1 reviewed (changes requested):
@fhnaseer currently if you do:

dotnet test --collect "Code Coverage;XPlat Code Coverage"

we are enabling both collectors and getting 2 reports from dynamic and coverlet so we have to keep this logic.
It means that we can support arguments only using such thing:

--collect "datacollector1;arg1=val1,arg2=val2"

and for more collectors:

--collect "datacollector1;arg1=val1,arg2=val2;datacollector2;arg3=val3,arg4=val4"

you probably need to be checking if = is in the collector name and then treat as argument to previous collector. Adding acceptance tests for both scenarios would be great.

fhnaseer committed:
Update existing configuration if present,

fhnaseer committed:
using correct tag for specifying output file format,

fhnaseer committed:
Merge remote-tracking branch 'origin/main' into dev/fhnaseer/cc-commandline-config

dotnet-maestro[bot] committed:
Update dependencies from https://dev.azure.com/devdiv/DevDiv/_git/vs-code-coverage build 20211111.1

Microsoft.Internal.CodeCoverage
 From Version 17.1.0-beta.21560.1 -> To Version 17.1.0-beta.21561.1

fhnaseer committed:
Merge remote-tracking branch 'origin/darc-main-f51f545b-6a53-4780-9c1d-8f377d0f6c72' into dev/fhnaseer/cc-commandline-config

fhnaseer committed:
fixing tests,

jakubch1 commented on test/Microsoft.TestPlatform.AcceptanceTests/TranslationLayerTests/CodeCoverageTests.cs:
```diff
@@ -414,7 +414,7 @@ private string GetCodeCoverageRunSettings(int cpuCount, bool useClrIeInstrumenta
                                                       <CLRIEInstrumentationNetCore>{useClrIeInstrumentationEngine}</CLRIEInstrumentationNetCore>
                                                       <CLRIEInstrumentationNetFramework>{useClrIeInstrumentationEngine}</CLRIEInstrumentationNetFramework>
-                                                      <CoverageFileOutputFormats>{outputFormat}</CoverageFileOutputFormats>
+                                                      <OutputFormat>{outputFormat}</OutputFormat>
```
Should be changed to Format

fhnaseer replied:
Fixed,

fhnaseer committed:
fixing parameter name,

fhnaseer committed:
Adding test for overriding format from settings file using collect command,
